### ISSUE TITLE
Provide a means of getting audio meta data / include in instructions in how to do so

### ISSUE THREAD

**Issue by zachsa** [bbc/peaks.js#246]
I would like to get the metadata of the audio, but didn't see in the readme if this is possible. I see that there is a 'player' key in the peaksjs object, but I get NaN for the duration.

Is it currently possible to get audio metadata?

**Comment by chrisn** [comment]
Peaks.js doesn't provide a complete wrapper for the HTMLMediaElement interface, and you can still access the underlying media element directly. Maybe it should? Feedback on this is welcome. (Note: see also #237.)

But, in the specific case of duration, `peaks.player.getDuration()` should work.

**Comment by zachsa** [comment]
thank you - @chrisn. `getDuration()` is exactly what I needed.

I don't know enough about the the HTMLMediaElement to really give reasonable feedback on whether Peaks should/shouldn't wrap it.

I actually looked for the `<audio/>` element in the DOM, but couldn't find it since I guess that it is replaced by the canvas element, and then is kept in memory reference instead? In any case, I was unable to select it via this code: `var elem = document.querySelector("audio");`. The MDN documentation mentions this is required to access metadata from the element itself.

Then looking at the `peaks.js` init object, I tried to access a `peaks.mediaElement` object. But there isn't one (I'm guessing there is a method to retrieve it such as `getDuration()`)?

In any case, it looks like there are some API functions that are not included in the readme? Is there a place where functions such as `getDuration()` are included?

**Comment by zachsa** [comment]
I think that wrapping the HTMLMediaInterface would lower the barrier for entry to using this library as a first time <audio/> user

**Comment by chrisn** [comment]
> I actually looked for the <audio/> element in the DOM, but couldn't find it since I guess that it is replaced by the canvas element, and then is kept in memory reference instead? In any case, I was unable to select it via this code: var elem = document.querySelector("audio");. The MDN documentation mentions this is required to access metadata from the element itself.

I'm slightly confused by this, as `mediaElement` is a required parameter to `Peaks.init()`, so you need a reference to the media element to be able to construct a `Peaks` instance to start with. The `<audio>` element should remain in the DOM, Peaks.js won't replace it with a `<canvas>` element.

**Comment by zachsa** [comment]
@chrisn - I specify this HTML:

```
<div id="peaks-js-container">
  <audio controls id="audio">
    <source src="/test2-converted.mp3" type="audio/mpeg"/>
    This browser does not support audio. Please use a recent version of Chrome
  </audio>
</div>
```

Then I init l `Peaks.js` in the `componentDidMount` callback:

```
componentDidMount() {
    this.state.peaks = Peaks.init({
      container: document.querySelector('#peaks-js-container'),
      mediaElement: document.querySelector('audio'),
      ...
```

And when the waveform is rendered I get this:

```
<div id="peaks-js-container">
   <div class="waveform">
      <div class="zoom-container">
         <div class="konvajs-content" role="presentation" style="position: relative; width: 1136px; height: 125px;">
            ... 4 <canvas></canvas> elements
         </div>
      </div>
      <div class="overview-container">
         <div class="konvajs-content" role="presentation" style="position: relative; width: 1136px; height: 125px;">
            ... 5 <canvas></canvas> elements
         </div>
      </div>
   </div>
</div>
```

It definitely looks like the `<audio/>` element is being replaced since the IDs of the container div are the same. But in the `init` function I'm specifying `mediaElement` as element inside of the `container`. Is this not correct?

**Comment by chrisn** [comment]
Ah, so when I said Peaks.js won't replace the `<audio>` element, I wasn't expecting it to be inside the container `<div>`, something more like this:
```html
<div>
  <audio id="audio" controls="controls">
    <source src="https://example.com/audio.mp3" type="audio/mpeg"/>
  </audio>

  <div id="waveform-container"></div>
</div>
```
I think this issue has come up before. Maybe we need better documentation, or Peaks.js shouldn't simply replace the container element with its own content and instead add the elements it needs as children without removing existing content?

**Comment by zachsa** [comment]
I see that it's specified already in the documentation that the waveform container is separate to the audio tag. I missed that, and it still worked (until I wanted to actually find the audio element).

Thanks!

**Comment by chrisn** [comment]
I have added a note about leaving the container elements empty, as part of #258. (I'd be interested in your feedback on the changes in that PR.) Is there anything more we need to change, to be able to close this issue?

**Comment by zachsa** [comment]
I'm not sure what you mean WRT to containers @chrisn ? 

This issue can be closed. The problem I was having (the reason that I opened it) was that I was specifying an audio element inside of the div that peaks.js was adding it's DOM stuff to. So the audio element was being destroyed. I'm not sure why it was still working actually.

So when I specified the audio element separately I had full access to it using standard DOM interactions.

What I found worked best was retrieving audio via an XHR request, and then creating an audio element on the request completion. For some reason specifying `src=""` directly in an audio element resulted in weird bugs where audio files were retrieved in parts multiple times. Also, by using the XHR request I was able to use authentication on the endpoint that retrieved the audio (which was piped from another server).
